### ISSUE TITLE
[test] OpInfo gap: allclose-based compile-vs-eager checks are blind to signed-zero and signed-infinity divergences

### ISSUE THREAD

**Issue by raravind007** [pytorch/pytorch#188680]
# [test] OpInfo gap: allclose-based compile-vs-eager checks are blind to signed-zero and signed-infinity divergences

## Summary

The existing compile-vs-eager correctness check in `TestCommon` uses `torch.allclose`
(or `torch.testing.assert_close`) for output comparison.  IEEE 754 defines
`+0.0 == -0.0`, so `allclose` treats them as identical — meaning a compiler
pass that silently normalises `−0.0 → +0.0` (or flips the sign of an infinity)
will pass every existing correctness test even though it produced the wrong result.

This issue proposes adding a lightweight, opt-in parity check for
`signbit` / `isinf` / `isnan` to the OpInfo test framework so those
divergences can be caught automatically.

---

## Concrete example of the blind spot

```python
import torch

# IEEE 754: +0.0 and -0.0 compare equal
a = torch.tensor(-0.0)
b = torch.tensor(+0.0)

torch.allclose(a, b)           # True  — the sign disappears silently
torch.equal(a, b)              # True  — same
torch.signbit(a).item()        # True  (negative zero)
torch.signbit(b).item()        # False (positive zero)

# The sign of a zero input propagates to the sign of the resulting infinity.
# A pass that normalises -0 → +0 before reciprocal will silently flip the output:
torch.reciprocal(a)            # tensor(-inf)   ← correct
torch.reciprocal(b)            # tensor(inf)    ← wrong, but looks "close"

# For copysign the sign IS the result:
torch.copysign(torch.tensor(0.0), torch.tensor(-1.0))  # tensor(-0.)   eager
# If compiled returned tensor(0.) instead:
torch.allclose(
    torch.tensor(-0.0),
    torch.tensor(+0.0),
)  # True — bug is invisible to the current test
```

If `torch.compile` produced `+0.0` where eager produces `−0.0` for
`torch.copysign`, every existing test would pass. The sign difference would
only be caught by a `signbit`-aware assertion.

---

## Why this matters

Ops whose output sign is semantically load-bearing include (non-exhaustively):

| Op | Sign-sensitive case |
|---|---|
| `torch.copysign` | The sign bit is the entire output |
| `torch.atan2` | IEEE 754 §9.2.1 specifies 4 distinct signed-zero corner cases for boundary inputs |
| `torch.reciprocal` | `1/−0.0 = −inf`, `1/+0.0 = +inf` — sign of zero propagates to sign of inf |
| `torch.neg` | `neg(−0.0) = +0.0` |
| `torch.sign` | `sign(−0.0)` is implementation-defined |
| `torch.pow`, `torch.div` | Various IEEE 754 signed-zero propagation rules |

A compiler optimisation that canonicalises `−0.0 → +0.0` in any of these
will produce silently wrong results that `allclose` cannot detect.

### Related issues

- #53963 — `torch.signbit` itself is currently incorrect for negative zero
  in eager, which means *both* sides of the comparison may be wrong. The
  proposed test infrastructure would surface those cases too.
- #172529 — `torch.isin` eager/compile divergence (shape mismatch), found
  without automated special-value coverage. Demonstrates that compile-vs-eager
  divergences in output structure are already slipping through.
- #164186 — Fuzzer-discovered eager/compile divergence (compile raises,
  eager succeeds). The exception-parity variant of this problem.

---

## Proposed change

Add a single opt-in boolean field to `OpInfo`:

```python
# torch/testing/_internal/opinfo/core.py
check_sign_sensitive_special_values: bool = False
```

And a companion test in `TestCommon` in `test/test_ops.py`:

```python
@ops(op_db, allowed_dtypes=floating_and_complex_types())
def test_sign_sensitive_special_values(self, device, dtype, op):
    """
    Checks that signbit / isinf / isnan are identical between
    eager and torch.compile on the op's reference_inputs.
    """
```

The test:
1. Skips unless `check_sign_sensitive_special_values=True` on the OpInfo
2. Prefers `reference_inputs` (which already generates ±inf, ±0.0 extremal
   combinations) over `sample_inputs`
3. Checks three properties in order:
   - **isnan parity**: NaN positions match
   - **isinf parity**: inf positions match
   - **signbit parity**: sign of every ±inf and every ±0.0 matches
4. Fails with a clear message showing position, eager value, and compiled value

The initial PR will enable the flag on three seed ops (`copysign`, `atan2`,
`reciprocal`) to prove the mechanism works end-to-end. Additional ops can be
opted in via follow-up PRs.

---

## What this does NOT include (intentional scope limit)

- Exception parity testing (eager raises, compiled succeeds silently) — separate issue/PR
- NaN payload comparison — out of scope, PyTorch does not guarantee NaN payload propagation
- Non-floating-point types — signed zeros and signed infinities are a floating-point concept

---

## cc

`@pytorch/pt2` for the compile-vs-eager aspect

cc @mruberry @chauhang @penguinwu

**Comment by raravind007** [comment]
Could a maintainer mark this issue as actionable if the proposed direction looks reasonable? I’ll wait for that before opening a PR.

**Comment by morrison-turnansky** [comment]
Generally, we express correctness within a tolerance value of eager. However, we have some issues/prs enforcing this specific exact correctness case. 

Do we have a subset of operations that we want to enforce this on? I think enforcing exact correctness in general is not good goal and will not be maintainable in the long run.

**Comment by raravind007** [comment]
Thanks you Morrison, that makes sense. I agree that enforcing exact correctness generally would not be maintainable, and that is not what I'm proposing.

The narrower scope I had in mind is an opt-in check for operations where special-value semantics are directly observable and tolerance-based comparison can miss a real behavioral difference.

For example, not exact elementwise equality in general, but targeted checks like:

- NaN mask matches
- Inf mask matches
- signbit matches only where the output is zero, and only for ops where signed zero is part of the observable behavior

A reasonable initial subset could be very small, for example ops where signed zero directly affects downstream results:
- torch.copysign
- torch.atan2
- possibly torch.reciprocal / division-style ops where +0.0 vs -0.0 can become +inf vs -inf

I'm happy to keep the first PR limited to adding the helper/opt-in mechanism and enabling it only for 2-3 seed ops, rather than applying it broadly across OpInfo. The goal would be to create a targeted guardrail for semantic special-value behavior, not to make exact correctness the default expectation.

**Comment by morrison-turnansky** [comment]
yes just wanted to clarify scope and give context to what "correct" means here. can you post a draft pr?
